variable definition bug

```php
var config = ["aaa":"bbb"];
var conf   = config["aaa"];
```
output:
```
StaticTypeInference=array-access undefined
StaticTypeInference=array-access variable
StaticTypeInference=array-access undefined
StaticTypeInference=array-access variable
```

```php
var config = ["aaa":"bbb"];
var aaaa   = config["aaa"];
// resumed normal
```
---
```php
var load = new Loader;
```
output:
```
 [ERROR] Unknown type: new
```

```php
var ddd = new Loader;
// resumed normal
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@ and this project adheres to [Semantic Versioning](https://semver.org).
 - **BC**: methods declared with a multi-type union return now carry an enforced return type in the compiled extension. PHP subclasses that override such a method must declare a compatible (equal or narrower) return type, where previously the missing return type made any override valid [#2428](https://github.com/zephir-lang/zephir/issues/2428)
 
 ### Fixed
+- Fixed the static type-inference pass leaking debug output (`StaticTypeInference=...` / `STI=...`) to stdout during compilation when it met a variable type or expression node it did not enumerate (e.g. reassigning a var holding a `(object)` cast). Such cases now silently degrade the variable to dynamic [#1877](https://github.com/zephir-lang/zephir/issues/1877)
 - Fixed concat-assign (`let this->prop .= value`) on object properties overwriting the property instead of appending to it; the current value is now read and concatenated [#2063](https://github.com/zephir-lang/zephir/issues/2063)
 - Fixed a false `Reached end of the method without returning a valid type` error for a method ending in an exhaustive `switch` (one with a `default` clause that returns/throws on every path) [#1706](https://github.com/zephir-lang/zephir/issues/1706)
 - Fixed `char`/`uchar` return values sign-extending when the byte exceeds 127 (e.g. `0x8A` returned `-118` instead of `138`). Char returns are now cast to `unsigned char` to preserve byte semantics [#1291](https://github.com/zephir-lang/zephir/issues/1291)

--- a/ext/stub/arrayaccesstest.zep.c
+++ b/ext/stub/arrayaccesstest.zep.c
@@ -141,6 +141,31 @@ PHP_METHOD(Stub_ArrayAccessTest, issue645)
 }
 
 /**
+ * Declaring a variable from an array-access on a hash literal. This is the
+ * original snippet reported in the issue; the static type-inference pass
+ * used to leak debug output to stdout while compiling it.
+ *
+ * @issue https://github.com/zephir-lang/zephir/issues/1877
+ */
+PHP_METHOD(Stub_ArrayAccessTest, issue1877)
+{
+	zval config, conf;
+	zephir_method_globals *ZEPHIR_METHOD_GLOBALS_PTR = NULL;
+
+	ZVAL_UNDEF(&config);
+	ZVAL_UNDEF(&conf);
+	ZEPHIR_METHOD_GLOBALS_PTR = pecalloc(1, sizeof(zephir_method_globals), 0);
+	zephir_memory_grow_stack(ZEPHIR_METHOD_GLOBALS_PTR, __func__);
+
+	ZEPHIR_INIT_VAR(&config);
+	zephir_create_array(&config, 1, 0);
+	add_assoc_stringl_ex(&config, SL("aaa"), SL("bbb"));
+	zephir_memory_observe(&conf);
+	zephir_array_fetch_string(&conf, &config, SL("aaa"), PH_NOISY, "stub/arrayaccesstest.zep", 70);
+	RETURN_CCTOR(&conf);
+}
+
+/**
  * @issue https://github.com/zephir-lang/zephir/issues/1155
  */
 PHP_METHOD(Stub_ArrayAccessTest, issue1155)
@@ -362,9 +387,9 @@ PHP_METHOD(Stub_ArrayAccessTest, issue1259UnsetKeyFromArrayInternalVariable)
 	zephir_create_array(&unsetData, 2, 0);
 	add_assoc_stringl_ex(&unsetData, SL("key_a"), SL("marcin"));
 	add_assoc_stringl_ex(&unsetData, SL("key_b"), SL("paula"));
-	zephir_array_append(&ret, &unsetData, PH_SEPARATE, "stub/arrayaccesstest.zep", 154);
+	zephir_array_append(&ret, &unsetData, PH_SEPARATE, "stub/arrayaccesstest.zep", 171);
 	zephir_array_unset_string(&unsetData, SL("key_a"), PH_SEPARATE);
-	zephir_array_append(&ret, &unsetData, PH_SEPARATE, "stub/arrayaccesstest.zep", 156);
+	zephir_array_append(&ret, &unsetData, PH_SEPARATE, "stub/arrayaccesstest.zep", 173);
 	RETURN_CTOR(&ret);
 }
 
@@ -389,14 +414,14 @@ PHP_METHOD(Stub_ArrayAccessTest, issue1259UnsetStringKeyFromArrayProperty)
 	ZEPHIR_INIT_VAR(&ret);
 	array_init(&ret);
 	zephir_read_property(&_0, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
-	zephir_array_append(&ret, &_0, PH_SEPARATE, "stub/arrayaccesstest.zep", 168);
+	zephir_array_append(&ret, &_0, PH_SEPARATE, "stub/arrayaccesstest.zep", 185);
 	zephir_read_property(&_1, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
 	ZVAL_STRING(&_1, "key_a");
 	zephir_unset_property_array(this_ptr, ZEND_STRL("unsetData"), &_1);
 	zephir_read_property(&_2, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
 	zephir_array_unset_string(&_2, SL("key_a"), PH_SEPARATE);
 	zephir_read_property(&_3, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
-	zephir_array_append(&ret, &_3, PH_SEPARATE, "stub/arrayaccesstest.zep", 170);
+	zephir_array_append(&ret, &_3, PH_SEPARATE, "stub/arrayaccesstest.zep", 187);
 	RETURN_CTOR(&ret);
 }
 
@@ -418,14 +443,14 @@ PHP_METHOD(Stub_ArrayAccessTest, issue1259UnsetLongKeyFromArrayProperty)
 	ZEPHIR_INIT_VAR(&ret);
 	array_init(&ret);
 	zephir_read_property(&_0, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
-	zephir_array_append(&ret, &_0, PH_SEPARATE, "stub/arrayaccesstest.zep", 179);
+	zephir_array_append(&ret, &_0, PH_SEPARATE, "stub/arrayaccesstest.zep", 196);
 	zephir_read_property(&_1, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
 	ZVAL_LONG(&_1, 3);
 	zephir_unset_property_array(this_ptr, ZEND_STRL("unsetData"), &_1);
 	zephir_read_property(&_2, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
 	zephir_array_unset_long(&_2, 3, PH_SEPARATE);
 	zephir_read_property(&_3, this_ptr, ZEND_STRL("unsetData"), PH_NOISY_CC | PH_READONLY);
-	zephir_array_append(&ret, &_3, PH_SEPARATE, "stub/arrayaccesstest.zep", 181);
+	zephir_array_append(&ret, &_3, PH_SEPARATE, "stub/arrayaccesstest.zep", 198);
 	RETURN_CTOR(&ret);
 }
 

--- a/ext/stub/arrayaccesstest.zep.h
+++ b/ext/stub/arrayaccesstest.zep.h
@@ -8,6 +8,7 @@ PHP_METHOD(Stub_ArrayAccessTest, get);
 PHP_METHOD(Stub_ArrayAccessTest, unsetByKeyFromArray);
 PHP_METHOD(Stub_ArrayAccessTest, unsetByKeyFromProperty);
 PHP_METHOD(Stub_ArrayAccessTest, issue645);
+PHP_METHOD(Stub_ArrayAccessTest, issue1877);
 PHP_METHOD(Stub_ArrayAccessTest, issue1155);
 PHP_METHOD(Stub_ArrayAccessTest, issue1094Test1);
 PHP_METHOD(Stub_ArrayAccessTest, issue1094Test2);
@@ -38,6 +39,9 @@ ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_stub_arrayaccesstest_unsetbykeyf
 ZEND_END_ARG_INFO()
 
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_stub_arrayaccesstest_issue645, 0, 0, IS_ARRAY, 0)
+ZEND_END_ARG_INFO()
+
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_stub_arrayaccesstest_issue1877, 0, 0, IS_STRING, 0)
 ZEND_END_ARG_INFO()
 
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_stub_arrayaccesstest_issue1155, 0, 0, _IS_BOOL, 0)
@@ -87,6 +91,7 @@ PHP_ME(Stub_ArrayAccessTest, get, arginfo_stub_arrayaccesstest_get, ZEND_ACC_PUB
 	PHP_ME(Stub_ArrayAccessTest, unsetByKeyFromArray, arginfo_stub_arrayaccesstest_unsetbykeyfromarray, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_ArrayAccessTest, unsetByKeyFromProperty, arginfo_stub_arrayaccesstest_unsetbykeyfromproperty, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_ArrayAccessTest, issue645, arginfo_stub_arrayaccesstest_issue645, ZEND_ACC_PUBLIC)
+	PHP_ME(Stub_ArrayAccessTest, issue1877, arginfo_stub_arrayaccesstest_issue1877, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_ArrayAccessTest, issue1155, arginfo_stub_arrayaccesstest_issue1155, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_ArrayAccessTest, issue1094Test1, arginfo_stub_arrayaccesstest_issue1094test1, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_ArrayAccessTest, issue1094Test2, arginfo_stub_arrayaccesstest_issue1094test2, ZEND_ACC_PUBLIC)

--- a/ext/stub/cast.zep.c
+++ b/ext/stub/cast.zep.c
@@ -1056,3 +1056,33 @@ PHP_METHOD(Stub_Cast, testIssue828)
 	RETURN_CTOR(&ret);
 }
 
+/**
+ * Reassigning a variable whose first inferred type comes from an
+ * `(object)` cast used to make the static type-inference pass leak
+ * `StaticTypeInference=object undefined` to stdout during compilation.
+ *
+ * @see https://github.com/zephir-lang/zephir/issues/1877
+ */
+PHP_METHOD(Stub_Cast, testObjectCastThenReassign)
+{
+	zval _0;
+	zval data, _1;
+	zephir_method_globals *ZEPHIR_METHOD_GLOBALS_PTR = NULL;
+
+	ZVAL_UNDEF(&data);
+	ZVAL_UNDEF(&_1);
+	ZVAL_UNDEF(&_0);
+	ZEPHIR_METHOD_GLOBALS_PTR = pecalloc(1, sizeof(zephir_method_globals), 0);
+	zephir_memory_grow_stack(ZEPHIR_METHOD_GLOBALS_PTR, __func__);
+
+	ZEPHIR_INIT_VAR(&_0);
+	zephir_create_array(&_0, 1, 0);
+	add_assoc_stringl_ex(&_0, SL("key"), SL("value"));
+	ZEPHIR_CPY_WRT(&_1, &_0);
+	zephir_convert_to_object(&_1);
+	ZEPHIR_CPY_WRT(&data, &_1);
+	ZEPHIR_INIT_NVAR(&data);
+	ZVAL_STRING(&data, "reassigned");
+	RETURN_CCTOR(&data);
+}
+

--- a/ext/stub/cast.zep.h
+++ b/ext/stub/cast.zep.h
@@ -75,6 +75,7 @@ PHP_METHOD(Stub_Cast, testArrayCastFromVariableFloat);
 PHP_METHOD(Stub_Cast, testArrayCastFromVariableString);
 PHP_METHOD(Stub_Cast, testArrayCastFromVariableStdClass);
 PHP_METHOD(Stub_Cast, testIssue828);
+PHP_METHOD(Stub_Cast, testObjectCastThenReassign);
 
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_stub_cast_testcharcastfromchar, 0, 0, IS_LONG, 0)
 ZEND_END_ARG_INFO()
@@ -294,6 +295,9 @@ ZEND_END_ARG_INFO()
 ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_stub_cast_testissue828, 0, 0, IS_ARRAY, 0)
 ZEND_END_ARG_INFO()
 
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_stub_cast_testobjectcastthenreassign, 0, 0, IS_STRING, 0)
+ZEND_END_ARG_INFO()
+
 ZEPHIR_INIT_FUNCS(stub_cast_method_entry) {
 	PHP_ME(Stub_Cast, testCharCastFromChar, arginfo_stub_cast_testcharcastfromchar, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_Cast, testCharCastFromVariableChar, arginfo_stub_cast_testcharcastfromvariablechar, ZEND_ACC_PUBLIC)
@@ -367,5 +371,6 @@ PHP_ME(Stub_Cast, testObjectCastFromString, arginfo_stub_cast_testobjectcastfrom
 	PHP_ME(Stub_Cast, testArrayCastFromVariableString, arginfo_stub_cast_testarraycastfromvariablestring, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_Cast, testArrayCastFromVariableStdClass, arginfo_stub_cast_testarraycastfromvariablestdclass, ZEND_ACC_PUBLIC)
 	PHP_ME(Stub_Cast, testIssue828, arginfo_stub_cast_testissue828, ZEND_ACC_PUBLIC)
+	PHP_ME(Stub_Cast, testObjectCastThenReassign, arginfo_stub_cast_testobjectcastthenreassign, ZEND_ACC_PUBLIC)
 	PHP_FE_END
 };

--- a/src/Passes/StaticTypeInference.php
+++ b/src/Passes/StaticTypeInference.php
@@ -17,8 +17,6 @@ use Zephir\StatementsBlock;
 
 use function is_string;
 
-use const PHP_EOL;
-
 /**
  * This pass try to infer typing on dynamic variables so the compiler
  * can replace them by low level types automatically
@@ -179,8 +177,9 @@ class StaticTypeInference
                 break;
 
             default:
-                // TODO: Find the reason
-                echo 'StaticTypeInference=', $currentType, ' ', $type, PHP_EOL;
+                // The current type is not one we can merge with; degrade the
+                // variable to a dynamic one so it is not optimized away.
+                $this->variables[$variable] = 'undefined';
                 break;
         }
     }
@@ -419,8 +418,8 @@ class StaticTypeInference
                 return 'undefined';
 
             default:
-                echo 'STI=', $expression['type'], PHP_EOL;
-                break;
+                // Unknown expression node: do not infer a type for it.
+                return null;
         }
     }
 

--- a/stub/arrayaccesstest.zep
+++ b/stub/arrayaccesstest.zep
@@ -56,6 +56,23 @@ class ArrayAccessTest
     }
 
     /**
+     * Declaring a variable from an array-access on a hash literal. This is the
+     * original snippet reported in the issue; the static type-inference pass
+     * used to leak debug output to stdout while compiling it.
+     *
+     * @issue https://github.com/zephir-lang/zephir/issues/1877
+     */
+    public function issue1877() -> string
+    {
+        var config, conf;
+
+        let config = ["aaa": "bbb"];
+        let conf   = config["aaa"];
+
+        return conf;
+    }
+
+    /**
      * @issue https://github.com/zephir-lang/zephir/issues/1155
      */
     public function issue1155() -> bool

--- a/stub/cast.zep
+++ b/stub/cast.zep
@@ -513,4 +513,21 @@ class Cast
 
         return ret;
     }
+
+    /**
+     * Reassigning a variable whose first inferred type comes from an
+     * `(object)` cast used to make the static type-inference pass leak
+     * `StaticTypeInference=object undefined` to stdout during compilation.
+     *
+     * @see https://github.com/zephir-lang/zephir/issues/1877
+     */
+    public function testObjectCastThenReassign() -> string
+    {
+        var data;
+
+        let data = (object) ["key": "value"];
+        let data = "reassigned";
+
+        return data;
+    }
 }

--- a/tests/Extension/ArrayAccessTest.php
+++ b/tests/Extension/ArrayAccessTest.php
@@ -34,6 +34,16 @@ final class ArrayAccessTest extends TestCase
         $this->assertSame(2, $class->get());
     }
 
+    /**
+     * @see https://github.com/zephir-lang/zephir/issues/1877
+     */
+    public function testIssue1877(): void
+    {
+        $class = new \Stub\ArrayAccessTest();
+
+        $this->assertSame('bbb', $class->issue1877());
+    }
+
     public function testUnsetByKeyFromArray(): void
     {
         $class = new \Stub\ArrayAccessTest();

--- a/tests/Extension/CastTest.php
+++ b/tests/Extension/CastTest.php
@@ -223,4 +223,12 @@ final class CastTest extends TestCase
         $this->assertSame(0.0, $return[4]);
         $this->assertSame(0, $return[5]);
     }
+
+    /**
+     * @see https://github.com/zephir-lang/zephir/issues/1877
+     */
+    public function testObjectCastThenReassign(): void
+    {
+        $this->assertSame('reassigned', $this->test->testObjectCastThenReassign());
+    }
 }

--- a/tests/Zephir/Passes/StaticTypeInferenceTest.php
+++ b/tests/Zephir/Passes/StaticTypeInferenceTest.php
@@ -85,4 +85,38 @@ final class StaticTypeInferenceTest extends TestCase
         // The pass method processes statements but returns null
         $this->assertNull($result);
     }
+
+    /**
+     * Reassigning a variable whose previously inferred type is not enumerated
+     * in the merge switch (e.g. it came from a `(object)` cast or an
+     * array-access) must not leak "StaticTypeInference=..." debug noise to
+     * stdout; the variable should instead degrade to a dynamic one.
+     *
+     * @see https://github.com/zephir-lang/zephir/issues/1877
+     */
+    public function testShouldNotEmitDebugOutputWhenMergingUnhandledType(): void
+    {
+        $this->expectOutputString('');
+
+        $this->inference->markVariable('x', 'object'); // e.g. from a (object) cast
+        $this->inference->markVariable('x', 'undefined'); // reassignment
+
+        // A conflicting reassignment downgrades to a dynamic variable.
+        $this->assertFalse($this->inference->getInferedType('x'));
+    }
+
+    /**
+     * The companion default branch in passExpression() must stay silent for
+     * expression node types it does not know about.
+     *
+     * @see https://github.com/zephir-lang/zephir/issues/1877
+     */
+    public function testShouldNotEmitDebugOutputForUnhandledExpressionType(): void
+    {
+        $this->expectOutputString('');
+
+        $result = $this->inference->passExpression(['type' => '__unknown__']);
+
+        $this->assertNull($result);
+    }
 }

